akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -129,64 +151,61 @@ private[remote] final class HeaderBuilderImpl(val compressionTable: LiteralCompr
   var _manifest: String = null
   var _manifestIdx: Int = -1
 
-  def senderActorRef_=(ref: String): Unit = {
-    _senderActorRef = ref
-    _senderActorRefIdx = compressionTable.compressActorRef(ref)
+  def senderActorRef_=(ref: ActorRef): Unit = {
+    _senderActorRefIdx = outboundCompression.compressActorRef(ref)
+    if (_senderActorRefIdx == -1) _senderActorRef = Serialization.serializedActorPath(ref) // includes local address from `currentTransportInformation`
   }
-
   def setNoSender(): Unit = {
     _senderActorRef = null
     _senderActorRefIdx = EnvelopeBuffer.DeadLettersCode
   }
-
   def isNoSender: Boolean =
     (_senderActorRef eq null) && _senderActorRefIdx == EnvelopeBuffer.DeadLettersCode
-
-  def senderActorRef: String = {
+  def senderActorRef: OptionVal[ActorRef] =
+    if (_senderActorRef eq null) inboundCompression.decompressActorRef(_senderActorRefIdx)
+    else OptionVal.None
+  def senderActorRefPath: String =
     if (_senderActorRef ne null) _senderActorRef
     else {
-      _senderActorRef = compressionTable.decompressActorRef(_senderActorRefIdx)
+      _senderActorRef = inboundCompression.decompressActorRef(_senderActorRefIdx).get.path.toSerializationFormat
       _senderActorRef
     }
-  }
 
   def setNoRecipient(): Unit = {
     _recipientActorRef = null
     _recipientActorRefIdx = EnvelopeBuffer.DeadLettersCode
   }
-
   def isNoRecipient: Boolean =
     (_recipientActorRef eq null) && _recipientActorRefIdx == EnvelopeBuffer.DeadLettersCode
 
-  def recipientActorRef_=(ref: String): Unit = {
-    _recipientActorRef = ref
-    _recipientActorRefIdx = compressionTable.compressActorRef(ref)
+  def recipientActorRef_=(ref: ActorRef): Unit = {
+    _recipientActorRefIdx = outboundCompression.compressActorRef(ref)
+    if (_recipientActorRefIdx == -1) _recipientActorRef = ref.path.toSerializationFormat
```

Comment:
I wonder what we should do with `ref.path.toSerializationFormat`. That is super slow and I think it might be worth caching it. Perhaps we have a cached field for it in `RemoteActorRef`?